crc32trim.c

```diff
@@ -62,64 +75,163 @@ static unsigned long crc32_trim_trailing_(
     if (len2 <= 0)
         return crc1;
 
-    /* get crcA0 */
-    crc1 ^= crc2;
+    if (operation == CRC_OPERATION_TRIM_TRAILING) {
+        /* put crcA0 in crc1 */
+        crc1 ^= crc2;
+    }
 
     /* put operator for one zero bit in odd */
-    odd[0] = 0xdb710641UL;          /* CRC-32 "Un"polynomial */
-    row = 0x80000000UL;
+    odd[0] = poly;
+    if (form == CRC_FORM_NORMAL) {
+        row = 0x80000000UL;
+    } else {
+        row = 1;
+    }
     for (n = 1; n < GF2_DIM; n++) {
         odd[n] = row;
-        row >>= 1;
+        if (form == CRC_FORM_NORMAL) {
+            row >>= 1;
+        } else {
+            row <<= 1;
+        }
     }
 
     /* put operator for two zero bits in even */
-    gf2_matrix_square(even, odd);
+    gf2_matrix_square(even, odd, form);
 
     /* put operator for four zero bits in odd */
-    gf2_matrix_square(odd, even);
+    gf2_matrix_square(odd, even, form);
 
     /* apply len2 zeros to crc1 (first square will put the operator for one
        zero byte, eight zero bits, in even) */
     do {
         /* apply zeros operator for this bit of len2 */
-        gf2_matrix_square(even, odd);
+        gf2_matrix_square(even, odd, form);
         if (len2 & 1)
-            crc1 = gf2_matrix_times(even, crc1);
+            crc1 = gf2_matrix_times(even, crc1, form);
         len2 >>= 1;
 
         /* if no more bits set, then done */
         if (len2 == 0)
             break;
 
         /* another iteration of the loop with odd and even swapped */
-        gf2_matrix_square(odd, even);
+        gf2_matrix_square(odd, even, form);
         if (len2 & 1)
-            crc1 = gf2_matrix_times(odd, crc1);
+            crc1 = gf2_matrix_times(odd, crc1, form);
         len2 >>= 1;
 
         /* if no more bits set, then done */
     } while (len2 != 0);
 
+    if (operation == CRC_OPERATION_COMBINE) {
+        /* put crcAB in crc1 */
+        crc1 ^= crc2;
+    }
+
     return crc1;
 }
 
-#ifndef NO_ZLIB
+/* ========================================================================= */
+unsigned long crc32_combine_nz(
+    unsigned long crcA,
+    unsigned long crcB,
+    long lenB
+) {
+    unsigned long poly = 0xedb88320UL; /* CRC-32 "reversed" polynomial */
+    return crc32_reconstruct(crcA, crcB, lenB,
+               poly, CRC_FORM_REFLECTED, CRC_OPERATION_COMBINE);
+}
+
 /* ========================================================================= */
 unsigned long crc32_trim_leading(
     unsigned long crcAB,
     unsigned long crcA,
     long lenB
 ) {
-    return crc32_combine(crcA, crcAB, lenB);
+    return crc32_combine_nz(crcA, crcAB, lenB);
 }
-#endif
 
 /* ========================================================================= */
 unsigned long crc32_trim_trailing(
     unsigned long crcAB,
     unsigned long crcB,
     long lenB
 ) {
-    return crc32_trim_trailing_(crcAB, crcB, lenB);
+    unsigned long poly = 0xdb710641UL; /* CRC-32 "reciprocal" polynomial */
```

Comment:
This poly seems to be from the original CRC-32 spec, but it looks like later on the polynomials being used are from the Castagnoli/Koopman stuff. But you do later on use the Castagnoli reciprocal polynomial (0x05EC76F1). Is there a reasoning for switching?

NOTE: I don't know anything about this space, and was just poking around to try to understand what all is happening here

Reply:
The combine and trim operations at the end of the day an xor and a matrix trick to quickly calculate the same result as feeding N null bytes into the crc state machine.

The xor let's us do one of the following.
```c
crcAB = crcA0 ^ crcB; // combine
crcA0 = crcAB ^ crcB; // trim trailing
crcA = crcA0 ^ crcA; // trim leading
```

For combine and trim leading the xor needs to happen after we calculate crcA0, after the matrix trick has been applied to crcA.

For trim trailing the xor needs to happen before the matrix trick so we have crcA0 and can then use the matrix trick to compute crcA.

The matrix trick is what allows us to efficiently go from crcA to crcA0 or from crcA0 to crcA. For a thorough explanation on how this matrix trick works, I suggest reading these excellent posts from @madler.
https://stackoverflow.com/questions/23122312/crc-calculation-of-a-mostly-static-data-stream/23126768#23126768
https://groups.google.com/d/msg/comp.compression/SHyr5bp5rtc/PP5-pmv9-9sJ

There is a relationship between the "reversed" and "reciprocal" polynomials (really polynomial representations) which allows for, when using the opposite xor placement and matrix trick form, going from crcA to crcA0 ("reversed") and from crcA0 to crcA ("reciprocal").

The same relationship exists for the "normal" and "reversed reciprocal" polynomials; going from crcA to crcA0 ("normal") and from crcA0 to crcA ("reversed reciprocal").

Each function triad (a set containing an associated combine, trim leading, and trim trailing operation) uses a pair of polynomial representations to get it's particular job done. The `crc32_*`, `crc32c_*`, and `cksum_*` functions all work on different types of crc32s, as you discovered.